### PR TITLE
Add support for function streaming log

### PR DESCRIPTION
**Pull request by Flanker32**
Add support for function streaming log

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogManager.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+
+public enum FunctionStreamingLogManager {
+    INSTANCE;
+
+    public static final String STREAMING_LOG_NOT_STARTED = "Streaming log is not started.";
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogManager.java:46]: private

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogManager.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            } catch (Throwable e) {
+                ApplicationManager.getApplication().invokeLater(() -> PluginUtil.displayErrorDialog(
+                        "Failed to start streaming log",
+                        e.getMessage()));
+            }
+        });
+    }
+
+    public void closeStreamingLog(Project project, String subscriptionId, String appName, String appId) {
+        ProgressManager.getInstance().run(new Task.Backgroundable(null, "Closing Streaming Log...", false) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogManager.java:83]: why not use DefaultLoader.getIdeHelper().runInBackground?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/function/StopFunctionStreamingLogsAction.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import com.microsoft.tooling.msservices.helpers.Name;
+import com.microsoft.tooling.msservices.serviceexplorer.NodeActionEvent;
+import com.microsoft.tooling.msservices.serviceexplorer.NodeActionListener;
+import com.microsoft.tooling.msservices.serviceexplorer.azure.function.FunctionNode;
+
+import static com.microsoft.azuretools.telemetry.TelemetryConstants.FUNCTION;
+import static com.microsoft.azuretools.telemetry.TelemetryConstants.STOP_STREAMING_LOG_FUNCTION_APP;
+
+@Name("Stop Streaming Logs")
+public class StopFunctionStreamingLogsAction extends NodeActionListener {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/function/StopFunctionStreamingLogsAction.java:39]: too many dup codes as StartFunctionStreamingLogsAction ,can use a shared base class AbstractFunctionStreamingLogAction

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/function/StopFunctionStreamingLogsAction.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        this.name = functionNode.getFunctionAppName();
+    }
+
+    @Override
+    protected void actionPerformed(NodeActionEvent nodeActionEvent) throws AzureCmdException {
+        EventUtil.executeWithLog(FUNCTION, STOP_STREAMING_LOG_FUNCTION_APP,
+            (operation) -> {
+                FunctionStreamingLogManager.INSTANCE.closeStreamingLog(project, subscriptionId, name, functionId);
+            },
+            (exception) -> {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/function/StopFunctionStreamingLogsAction.java:60]: remove useless brace (exception)

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogManager.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                if (!consoleView.isEnable()) {
+                    FunctionApp functionApp = AzureFunctionMvpModel.getInstance()
+                                                                   .getFunctionById(subscriptionId, appId);
+                    // Todo: Check Whether file logging is enabled and prompt user to enable it
+                    Observable<String> log = functionApp.streamAllLogsAsync();
+                    consoleView.startStreamingLog(log);
+                }
+                consoleViewMap.put(appId, consoleView);
+                StreamingLogsToolWindowManager.getInstance().showStreamingLogConsole(
+                        project, appId, String.format("%s-%s", appName, appName), consoleView);
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogManager.java:73]: why show appName-appName here ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogConsoleView.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    public void closeStreamingLog() {
+        if (isEnable()) {
+            subscription.unsubscribe();
+            printlnToConsole(STOP_LOG_STREAMING, SYSTEM_OUTPUT);
+        }
+    }
+
+    public boolean isEnable() {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogConsoleView.java:65]: isSubscribing

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/function/StopFunctionStreamingLogsAction.java (hunk 3 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/function/StopFunctionStreamingLogsAction.java:38]: Remove some unused parameters in `StopFunctionStreamingLogsAction`, and i think its a bit heavy to make an abstract parent for two simple action.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogManager.java (hunk 5 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogManager.java]: Added by mistake, removed

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogConsoleView.java (hunk 6 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/function/FunctionStreamingLogConsoleView.java:65]: This method is to tell whether console is enabled for logging, as we will always create subscription when start logging and close it when stop, so return the result of `subscription.isUnsubscribed()`, so i think `isEnable` is more clear as a method name

**Pull request merged**